src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+            } while (currentState != knownState);
+
+            // Check for last reader
+            if (isLastReader)
+            {
+                // Check for waiting writers
+                if ((knownState & LockStates.WaitingWritersMask) != 0)
+                {
+                    Debug.Assert((_state & LockStates.WriterSignaled) != 0);
+                    Debug.Assert(writerEvent != null);
+                    writerEvent.Set();
+                }
+                // Check for waiting readers
+                else if ((knownState & LockStates.WaitingReadersMask) != 0)
+                {
+                    Debug.Assert((_state & LockStates.ReaderSignaled) != 0);
+                    Debug.Assert(readerEvent != null);
+                    readerEvent.Set();
+                }
+                // Check for the need to release events
+                else if (cacheEvents)
+                {
+                    ReleaseEvents();
+                }
+            }
+
+            Debug.Assert(lockEntry.IsEmpty);
+        }
+
+        public void ReleaseWriterLock()
+        {
+            if (_writerID != GetCurrentThreadID())
+            {
+                throw GetNotOwnerException();
+            }
+
+            Debug.Assert((_state & LockStates.ReadersMask) == 0);
+            Debug.Assert((_state & LockStates.Writer) != 0);
+            Debug.Assert(_writerLevel != 0);
+
+            // Check for nested release
+            --_writerLevel;
+            if (_writerLevel != 0)
+            {
+                return;
+            }
+
+            // Not a writer any more
+            _writerID = InvalidThreadID;
+            bool cacheEvents;
+            ManualResetEvent readerEvent = null;
+            AutoResetEvent writerEvent = null;
+            int currentState = _state;
+            int knownState;
+            do
+            {
+                cacheEvents = false;
+                knownState = currentState;
+                int modifyState = -LockStates.Writer;
+
+                if ((knownState & LockStates.WaitingReadersMask) != 0)
+                {
+                    try
+                    {
+                        readerEvent = GetOrCreateReaderEvent();
+                    }
+                    catch (SystemException)
+                    {
+                        Helpers.Sleep(100);
+                        currentState = _state;
+                        knownState = 0;
+                        Debug.Assert(currentState != knownState);
+                        continue;
+                    }
+                    modifyState += LockStates.ReaderSignaled;
+                }
+                else if ((knownState & LockStates.WaitingWritersMask) != 0)
+                {
+                    try
+                    {
+                        writerEvent = GetOrCreateWriterEvent();
+                    }
+                    catch (SystemException)
```

Comment:
Since a WaitingWriters bit is set, the event would usually already be created (if the waiting writer that called AcquireWriterLock is already waiting on the event, it would have created the event). However, AcquireWriterLock adds WaitingWriter to the state before the event is created, so new calls to AcquireWriteLock that add themselves as a WaitingWriter but have not started waiting yet, would first attempt to create the event. In a low-resource situation, AcquireWriterLock would throw. Eventually, all such new waiting writers would throw, and the WaitingWriters bits would not be set anymore, breaking the loop.